### PR TITLE
KTOR-6501 Update API Docs for HttpClient

### PR DESCRIPTION
**Pull request by e5l**
[KTOR-6501](https://youtrack.jetbrains.com/issue/KTOR-6501) Add Actionalable KDocs for the HttpClient public API

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 1]
```diff
@@ -86,6 +1222,14 @@ public class HttpClient(
         this.manageEngine = manageEngine
     }
 
+    private suspend fun x() {
```
Comment (osipxd) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:1225]: Should it be deleted?

File: ktor-client/ktor-client-tests/build.gradle.kts [hunk 2]
```diff
@@ -59,6 +59,7 @@ kotlin.sourceSets {
         dependencies {
             api(project(":ktor-client:ktor-client-apache"))
             api(project(":ktor-client:ktor-client-apache5"))
+            api(project(":ktor-client:ktor-client-cio"))
```
Comment (osipxd) [inline ktor-client/ktor-client-tests/build.gradle.kts:62]: Seems not related to the docs

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt (hunk 1 above)
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:1225]: thanks, will drop

File: ktor-client/ktor-client-tests/build.gradle.kts (hunk 2 above)
Comment (e5l) [reply ktor-client/ktor-client-tests/build.gradle.kts:62]: will drop as well

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 3]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *
+ * For the request timeout:
+ * ```kotlin
+ * client.get("") {
+ *     timeout {
+ *         requestTimeoutMillis = 1000
+ *     }
+ * }
+ * ```
+ * See [Timeout](https://ktor.io/docs/client-timeout.html) for more information.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:164]:
```suggestion
 * For more information, see the [Timeout](https://ktor.io/docs/client-timeout.html) documentation.
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 4]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *             loadTokens {
+ *                 BearerTokens(accessToken = "your_access_token", refreshToken = "your_refresh_token")
+ *             }
+ *         }
+ *     }
+ * }
+ *
+ * val response = client.get("https://api.example.com/protected")
+ * ```
+ * Refer to [Client authentication](https://ktor.io/docs/client-auth.html) for more details.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:142]:
```suggestion
 * For more details, refer to [Client authentication](https://ktor.io/docs/client-auth.html).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 5]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ * // Or using submitForm
+ * val response = client.submitForm(
+ *     url = "https://example.com/submit",
+ *     formParameters = Parameters.build {
+ *         append("username", "user")
+ *         append("password", "pass")
+ *     }
+ * )
+ * ```
+ * More information is available at [Submitting form parameters](https://ktor.io/docs/client-requests.html#form_parameters).
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:123]:
```suggestion
 * For more information, see [Submitting form parameters](https://ktor.io/docs/client-requests.html#form_parameters).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 6]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *
+ * @Serializable
+ * data class MyResponseType(val someResponseData: String)
+ *
+ * val response: MyResponseType = client.post("https://api.example.com/data") {
+ *     contentType(ContentType.Application.Json)
+ *     setBody(MyRequestType(someData = "value"))
+ * }.body()
+ * ```
+ * See [Serializing JSON data](https://ktor.io/docs/client-serialization.html) for maven configuration and other details.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:101]:
```suggestion
 * For Maven configuration and other details, see [Serializing JSON data](https://ktor.io/docs/client-serialization.html).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 7]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ * Include headers in your request using the `headers` builder or the `header` function:
+ * ```kotlin
+ * val response = client.get("https://httpbin.org/bearer") {
+ *     headers {
+ *         append("Authorization", "Bearer your_token_here")
+ *         append("Accept", "application/json")
+ *     }
+ * }
+ * ```
+ * Learn more at [Adding headers to a request](https://ktor.io/docs/client-requests.html#headers).
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:69]:
```suggestion
 * To learn more, see [Adding headers to a request](https://ktor.io/docs/client-requests.html#headers).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 8]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ * To learn more on how to create and configure an [HttpClient] see the tutorial page:
+ * [Creating and configuring a client](https://ktor.io/docs/create-client.html).
+ *
+ * # Making API Requests
+ * For every HTTP method (GET, POST, PUT, etc.), there is a corresponding function:
+ * ```kotlin
+ * val response: HttpResponse = client.get("https://ktor.io/")
+ * val body = response.bodyAsText()
+ * ```
+ * See [Making HTTP requests](https://ktor.io/docs/client-requests.html) for more details.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:46]:
```suggestion
 * For more details, see [Making HTTP requests](https://ktor.io/docs/client-requests.html).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 9]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *      url = "https://example.com/upload",
+ *      formData = formData {
+ *          append("description", "File upload example")
+ *          append("file", {
+ *              File("path/to/file.txt").readChannel()
+ *          })
+ *      }
+ *  )
+ *
+ * See [Uploading data](https://ktor.io/docs/client-requests.html#upload_file) for details.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:194]:
```suggestion
 * For more details, see [Uploading data](https://ktor.io/docs/client-requests.html#upload_file).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 10]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ * ```kotlin
+ * client.webSocket("wss://echo.websocket.org") {
+ *     send(Frame.Text("Hello, WebSocket!"))
+ *     val frame = incoming.receive()
+ *     if (frame is Frame.Text) {
+ *         println("Received: ${frame.readText()}")
+ *     }
+ * }
+ * ```
+ * Learn more at [Client WebSockets](https://ktor.io/docs/client-websockets.html).
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:208]:
```suggestion
 * To learn more, see [Client WebSockets](https://ktor.io/docs/client-websockets.html).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 11]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *         validateResponse { response ->
+ *             val statusCode = response.status.value
+ *             when (statusCode) {
+ *                 in 300..399 -> error("Redirects are not allowed")
+ *             }
+ *         }
+ *     }
+ * }
+
+ * See [Error handling](https://ktor.io/docs/client-response-validation.html) for more information.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:223]:
```suggestion
 * For more information, see [Error handling](https://ktor.io/docs/client-response-validation.html).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 12]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ * ```kotlin
+ * val client = HttpClient() {
+ *     engine {
+ *         proxy = ProxyBuilder.http("http://proxy.example.com:8080")
+ *         // For a SOCKS proxy:
+ *         // proxy = ProxyBuilder.socks(host = "proxy.example.com", port = 1080)
+ *     }
+ * }
+ * ```
+ * See [Using a proxy](https://ktor.io/docs/client-proxy.html) for details.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:241]:
```suggestion
 * For more details, see [Using a proxy](https://ktor.io/docs/client-proxy.html).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 13]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *     while (true) {
+ *         for (event in incoming) {
+ *             println("Event from server:")
+ *             println(event)
+ *         }
+ *     }
+ * }
+ * ```
+ *
+ * Visit [Using SSE](https://ktor.io/docs/client-server-sent-events.html#install_plugin) to learn more.
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:296]:
```suggestion
 * For more information, see [Using SSE](https://ktor.io/docs/client-server-sent-events.html#install_plugin).
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 14]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
 ): HttpClient
 
 /**
- * Creates an asynchronous [HttpClient] with the specified [HttpClientEngineFactory] and optional [block] configuration.
- * Note that a specific platform may require a specific engine for processing requests.
- * You can learn more about available engines from [Engines](https://ktor.io/docs/http-client-engines.html).
+ * A multiplatform asynchronous HTTP client that allows you to make requests, handle responses,
+ * and extend its functionality with plugins such as authentication, JSON serialization, and more.
+ *
+ * # Creating client
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:320]:
```suggestion
 * # Creating a client
```

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 15]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ * val client = HttpClient {
+ *     install(HttpCookies) {
+ *         storage = AcceptAllCookiesStorage()
+ *     }
+ * }
+ *
+ * // Accessing cookies
+ * val cookies: List<Cookie> = client.cookies("https://example.com")
+ * ```
+ * Learn more at [Cookies](https://ktor.io/docs/client-cookies.html).
```
Comment (vnikolova) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:179]:
```suggestion
 * For more information, see the [Cookies](https://ktor.io/docs/client-cookies.html) documentation.
```

**Review by vnikolova — APPROVED** [review]
Looks good to me! I suggested some links  corrections to keep them consistent.

File: ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt [hunk 16]
```diff
@@ -19,20 +19,589 @@ import kotlinx.coroutines.*
...
+ *         append("Authorization", "Bearer your_token_here")
+ *         append("Accept", "application/json")
+ *     }
+ * }
+ * ```
+ * Learn more at [Adding headers to a request](https://ktor.io/docs/client-requests.html#headers).
+ *
+ * # JSON Serialization
+ * Add dependencies:
+ * - io.ktor:ktor-client-content-negotiation:3.+
```
[the remark this answers is not in the document]
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/HttpClient.kt:73]: This docs will be only available after 3.0 update, so we can skip it

**Comment by e5l** [comment]
@vnikolova, thank you for the suggestions! Applied
